File: common/src/main/java/com/jmsoftware/maf/common/domain/authcenter/security/UserPrincipal.java
package com.jmsoftware.maf.common.domain.authcenter.security;

import cn.hutool.core.util.StrUtil;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.jmsoftware.maf.common.domain.authcenter.permission.GetPermissionListByRoleIdListResponse;
import com.jmsoftware.maf.common.domain.authcenter.user.GetUserByLoginTokenResponse;
import com.jmsoftware.maf.common.domain.authcenter.user.UserStatus;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.val;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.io.Serial;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

/**
 * <h1>UserPrincipal</h1>
 * <p>Custom user details.</p>
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com
 * @date 2019-03-23 20:52
 **/
@Data
@NoArgsConstructor
@AllArgsConstructor
public class UserPrincipal implements UserDetails {
    @Serial
    private static final long serialVersionUID = -53353171692896501L;

    /**
     * Primary key
     */
    private Long id;
    /**
     * Username
     */
    private String username;
    /**
     * Email
     */
    private String email;
    /**
     * Phone
     */
    private String cellphone;
    /**
     * Password
     */
    @JsonIgnore
    private String password;
    /**
     * Nickname
     */
    private String fullName;
    /**
     * Birthday
     */
    private LocalDate birthday;
    /**
     * Gender
     */
    private String gender;
    /**
     * Status
     */
    private Byte status;
    /**
     * Create time
     */
    private LocalDateTime gmtCreated;
    /**
     * Modify time
     */
    private LocalDateTime gmtModified;
    /**
     * Roles that user has
     */
    private List<String> roles;
    /**
     * Authorities that user has
     */
    private Collection<? extends GrantedAuthority> authorities;

    /**
     * Create by username user principal.
     *
     * @param username the username
     * @return the user principal
     * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 12/24/2020 3:12 PM
     */
    public static UserPrincipal createByUsername(String username) {
        UserPrincipal userPrincipal = new UserPrincipal();
        userPrincipal.setUsername(username);
        return userPrincipal;
    }

    /**
     * Create user principal
     *
     * @param user           user
     * @param roleNameList   role name list
     * @param permissionList permission po list
     * @return user principal
     */
    public static UserPrincipal create(GetUserByLoginTokenResponse user, List<String> roleNameList,
                                       List<GetPermissionListByRoleIdListResponse.Permission> permissionList) {
        val permissions =
                Optional.ofNullable(permissionList).orElse(new LinkedList<>());
        val authorities = permissions.stream()
                .filter(permission -> StrUtil.isNotBlank(permission.getPermissionExpression()))
                .map(permission -> new SimpleGrantedAuthority(permission.getPermissionExpression()))
                .toList();

        return new UserPrincipal(user.getId(),
                                 user.getUsername(),
                                 user.getEmail(),
                                 user.getCellphone(),
                                 user.getPassword(),
                                 user.getFullName(),
                                 user.getBirthday(),
                                 user.getGender(),
                                 user.getStatus(),
                                 user.getCreatedTime(),
                                 user.getModifiedTime(),
                                 roleNameList,
                                 authorities);
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return this.authorities;
    }

    @Override
    public String getPassword() {
        return this.password;
    }

    @Override
    public String getUsername() {
        return this.username;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return Objects.equals(this.status, UserStatus.ENABLED.getValue());
    }
}

File: common/src/main/java/com/jmsoftware/maf/common/domain/authcenter/security/UserPrincipal.kt
package com.jmsoftware.maf.common.domain.authcenter.security

import cn.hutool.core.util.StrUtil
import com.fasterxml.jackson.annotation.JsonIgnore
import com.jmsoftware.maf.common.domain.authcenter.permission.GetPermissionListByRoleIdListResponse
import com.jmsoftware.maf.common.domain.authcenter.user.GetUserByLoginTokenResponse
import com.jmsoftware.maf.common.domain.authcenter.user.UserStatus
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import java.io.Serial
import java.time.LocalDate
import java.time.LocalDateTime

/**
 * # UserPrincipal
 *
 * Custom user details.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/16/22 7:26 PM
 */
class UserPrincipal : UserDetails {
    /**
     * Primary key
     */
    var id: Long? = null

    /**
     * Username
     */
    var usernameProperty: String? = null

    /**
     * Email
     */
    var email: String? = null

    /**
     * Phone
     */
    var cellphone: String? = null

    /**
     * Password
     */
    @JsonIgnore
    var passwordProperty: String? = null

    /**
     * Nickname
     */
    var fullName: String? = null

    /**
     * Birthday
     */
    var birthday: LocalDate? = null

    /**
     * Gender
     */
    var gender: String? = null

    /**
     * Status
     */
    var status: Byte? = null

    /**
     * Create time
     */
    var gmtCreated: LocalDateTime? = null

    /**
     * Modify time
     */
    var gmtModified: LocalDateTime? = null

    /**
     * Roles that user has
     */
    var roles: List<String>? = null

    /**
     * Authorities that user has
     */
    var authoritiesProperty: Collection<GrantedAuthority>? = null

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return authoritiesProperty!!
    }

    override fun getPassword(): String {
        return passwordProperty!!
    }

    override fun getUsername(): String {
        return usernameProperty!!
    }

    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return status == UserStatus.ENABLED.value
    }

    companion object {
        @Serial
        private const val serialVersionUID = -53353171692896501L

        /**
         * Create by username user principal.
         *
         * @param username the username
         * @return the user principal
         * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 12/24/2020 3:12 PM
         */
        fun createByUsername(username: String): UserPrincipal {
            val userPrincipal = UserPrincipal()
            userPrincipal.usernameProperty = username
            return userPrincipal
        }

        /**
         * Create user principal
         *
         * @param user           user
         * @param roleNameList   role name list
         * @param permissionList permission po list
         * @return user principal
         */
        fun create(
            user: GetUserByLoginTokenResponse,
            roleNameList: List<String> = listOf(),
            permissionList: List<GetPermissionListByRoleIdListResponse.Permission> = listOf()
        ): UserPrincipal {
            val authorities = permissionList.stream()
                .filter { permission: GetPermissionListByRoleIdListResponse.Permission -> StrUtil.isNotBlank(permission.permissionExpression) }
                .map { permission: GetPermissionListByRoleIdListResponse.Permission -> SimpleGrantedAuthority(permission.permissionExpression) }
                .toList()
            return UserPrincipal().apply {
                this.id = user.id
                this.usernameProperty = user.username
                this.email = user.email
                this.cellphone = user.cellphone
                this.passwordProperty = user.password
                this.fullName = user.fullName
                this.birthday = user.birthday
                this.gender = user.gender
                this.status = user.status
                this.gmtCreated = user.createdTime
                this.gmtModified = user.modifiedTime
                this.roles = roleNameList
                this.authoritiesProperty = authorities
            }
        }
    }
}
